### ISSUE TITLE
Using $this is final class where static is expected leads to false-positive?

### ISSUE THREAD

**Issue by calvinalkan** [vimeo/psalm#7644]
I'm not 100% sure if this is a false positive but using $this inside a final class where static is expected should not trigger an error.

https://psalm.dev/r/868f28412c

```php
<?php

interface FooInterface {
 
    /**
    * @param Closure(static) $closure
    */
    public function doStuff(Closure $closure) :void;
    
    public function someMethod() :void;
    
}

final class Foo implements FooInterface{
    
    
    public function doStuff(Closure $closure) :void {
        
        // What to do here?
        $closure($this);
        
    }
    
    public function someMethod ():void {
    }
    
}

final class Bar implements FooInterface {

 public function doStuff(Closure $closure) :void {
        
     // What to do here?
        $closure($this);
        
    }
    
    public function someMethod ():void {
    }
}

$foo = new Foo();
// This is valid
$foo->doStuff(function(Foo $foo) {
    $foo->someMethod();
});

$bar = new Bar();
// This should fail and does.
$bar->doStuff(function(Foo $foo) {
    $foo->someMethod();
});
```

**Comment by orklah** [comment]
static as a param type for methods is a pretty weird concept

But static as a param type for closure is next level weird 😄 Given that the place where you define the closure and the place from where you'll call it could be completely different context, I can't even start to think about how this should behave

Psalm seems to support that a little, I'm not sure it was on purpose. I'm not sure it's a sane approach anyway.

I'll flag this as enhancement because I don't know if it would be right to reject it, but I don't think anyone will work on that anytime soon...
